external-crates/move/crates/move-compiler/src/linters/unneeded_return.rs

```diff
@@ -0,0 +1,216 @@
+        T::UnannotatedExp_::Return(rhs) => {
+            if returnable_value(context, rhs) {
+                report_unneeded_return(context, exp.exp.loc);
+            };
+            false
+        }
+
+        T::UnannotatedExp_::BinopExp(lhs, _, _, rhs) => {
+            returnable_value(context, lhs) && returnable_value(context, rhs)
+        }
+        T::UnannotatedExp_::ExpList(values) => values.iter().all(|v| match v {
+            T::ExpListItem::Single(exp, _) => returnable_value(context, exp),
+            T::ExpListItem::Splat(_, _, _) => false,
+        }),
+        T::UnannotatedExp_::Borrow(_, exp, _)
+        | T::UnannotatedExp_::Dereference(exp)
+        | T::UnannotatedExp_::UnaryExp(_, exp)
+        | T::UnannotatedExp_::TempBorrow(_, exp)
+        | T::UnannotatedExp_::Cast(exp, _)
+        | T::UnannotatedExp_::Annotate(exp, _) => returnable_value(context, exp),
+
+        T::UnannotatedExp_::Pack(_, _, _, _)
+        | T::UnannotatedExp_::PackVariant(_, _, _, _, _)
+        | T::UnannotatedExp_::Unit { .. }
+        | T::UnannotatedExp_::Value(_)
+        | T::UnannotatedExp_::Move { .. }
+        | T::UnannotatedExp_::Copy { .. }
+        | T::UnannotatedExp_::Use(_)
+        | T::UnannotatedExp_::Constant(_, _)
+        | T::UnannotatedExp_::ModuleCall(_)
+        | T::UnannotatedExp_::Builtin(_, _)
+        | T::UnannotatedExp_::BorrowLocal(_, _)
+        | T::UnannotatedExp_::ErrorConstant { .. }
+        | T::UnannotatedExp_::Vector(_, _, _, _) => true,
+
+        T::UnannotatedExp_::IfElse(_, _, _)
+        | T::UnannotatedExp_::Match(_, _)
+        | T::UnannotatedExp_::VariantMatch(_, _, _) => true,
+
+        // While loops can't yield values, so there should already be other errors.
+        T::UnannotatedExp_::While(_, _, _) => false,
+
+        // Non-while loops _can_ yield values, and should never appear after a return if the
+        // value is intended.
+        T::UnannotatedExp_::Loop { .. } => true,
+
+        T::UnannotatedExp_::NamedBlock(_, (_, seq)) | T::UnannotatedExp_::Block((_, seq)) => {
+            let Some(last) = seq.back() else { return false };
+            match &last.value {
+                T::SequenceItem_::Seq(exp) => returnable_value(context, exp),
+                T::SequenceItem_::Declare(_) | T::SequenceItem_::Bind(_, _, _) => false,
+            }
+        }
+
+        // These don't really make sense in return position, so there should already be other
+        // errors.
+        T::UnannotatedExp_::Assign(_, _, _)
+        | T::UnannotatedExp_::Mutate(_, _)
+        | T::UnannotatedExp_::Abort(_)
+        | T::UnannotatedExp_::Give(_, _)
+        | T::UnannotatedExp_::Continue(_) => false,
+        T::UnannotatedExp_::UnresolvedError => false,
+    }
+}
+
+fn report_unneeded_return(context: &mut Context, loc: Loc) {
+    context.env.add_diag(diag!(
+        StyleCodes::UnneededReturn.diag_info(),
+        (loc, "Remove 'return' from this statement")
```

Comment:
We should probably add some more context to this error. Since whoever wrote this code probably thought it was necessary. Maybe something like (just a first stab here)
```suggestion
        (loc, "Remove unnecessary 'return', the expression is already in a 'return' position")
```